sys/net/application_layer/nanocoap/sock.c

```diff
@@ -294,38 +308,33 @@ static int _block_cb(void *arg, coap_pkt_t *pkt)
     /* response was not block-wise */
     if (!coap_get_block2(pkt, &block2)) {
         block2.offset = 0;
-        ctx->more = false;
-    } else {
-        ctx->more = block2.more;
+        block2.more = false;
     }
 
+    ctx->more = block2.more > 0;
     return ctx->callback(ctx->arg, block2.offset, pkt->payload, pkt->payload_len, block2.more);
 }
 
-static int _fetch_block(coap_pkt_t *pkt, sock_udp_t *sock,
+static int _fetch_block(nanocoap_sock_t *sock, uint8_t *buf, size_t len,
                         const char *path, coap_blksize_t blksize, unsigned num,
                         _block_ctx_t *ctx)
 {
-    uint8_t *pktpos = (void *)pkt->hdr;
+    coap_pkt_t pkt = {
+        .hdr = (void *)buf,
+    };
     uint16_t lastonum = 0;
 
-    pktpos += coap_build_hdr(pkt->hdr, COAP_TYPE_CON, NULL, 0, COAP_METHOD_GET,
-                             num);
-    pktpos += coap_opt_put_uri_pathquery(pktpos, &lastonum, path);
-    pktpos += coap_opt_put_uint(pktpos, lastonum, COAP_OPT_BLOCK2,
-                                (num << 4) | blksize);
+    buf += coap_build_hdr(pkt.hdr, COAP_TYPE_CON, NULL, 0, COAP_METHOD_GET, _get_id());
```

Comment:
Seems like we should actually be setting the token (which is 0 right now) to maybe `num` and then check on that instead of `else if (coap_get_id(pkt) != id)`, would that be correct @maribu?

Reply:
Note that a zero length token is still a valid token. The standard asks for 32 bit of entropy in the Token if a device is connected to the Internet, though:

> A client that is connected to the general Internet SHOULD use at least 32 bits of randomness, keeping in mind that not being directly connected to the Internet is not necessarily sufficient protection against spoofing.  (Note that the Message ID adds little in protection as it is usually sequentially assigned, i.e., guessable, and can be circumvented by spoofing a separate response.)

https://datatracker.ietf.org/doc/html/rfc7252#section-5.3.1

Reply:
> Note that a zero length token is still a valid token

Yep

> The standard asks for 32 bit of entropy in the Token if a device is connected to the Internet, though

In that case, the default token length should at least be 4bytes, thinking of gcoap default https://github.com/RIOT-OS/RIOT/blob/e64b1bd7648a13c0d7530aab364391f8c54a2c5c/sys/include/net/gcoap.h#L482-L490

Although for this specific use-case, I guess if a block option is set it could check that the blknum matches what is expected, keeping the token field to 0.

Reply:
Do the individual blocks require different tokens or is this considered a single transfer, so the same token for all blocks?

Reply:
It would be different tokens for each block request, but I think you could leave it empty and check in the `nanocoap_sock_request_cb` call `coap_get_block2(pkt, &block2);` on the request and response, and if its there is a block option and the `blknum` does not match retry reception. But this optimization is maybe a bit of a hack to save the token bytes...

Reply:
It is up to the client and needs to aid the client to identify the right response. With NSTART=1 only duplicates (trivial to detect with blockwise) and spoofed responses are needed to be told apart from the right one. In case of UDP transport, one cannot really defend against spoofed responses, but having same entropy in the Token at least requires the adversary to be able to listen to the requests. The standard specifically says 32 bits of randomness and not reusing Tokens for the same source and destination endpoints is what SHOULD be done.

The common strategy is to just use a properly seeded PRNG for generating the token and hope for the best.

Reply:
Since we don't include a token yet in the request, how about we move this to a follow-up PR?